// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: data_message.proto

package go_protocol

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on DataMessageBody with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *DataMessageBody) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataMessageBody with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DataMessageBodyMultiError, or nil if none found.
func (m *DataMessageBody) ValidateAll() error {
	return m.validate(true)
}

func (m *DataMessageBody) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DataMessageBodyMultiError(errors)
	}

	return nil
}

// DataMessageBodyMultiError is an error wrapping multiple validation errors
// returned by DataMessageBody.ValidateAll() if the designated constraints
// aren't met.
type DataMessageBodyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataMessageBodyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataMessageBodyMultiError) AllErrors() []error { return m }

// DataMessageBodyValidationError is the validation error returned by
// DataMessageBody.Validate if the designated constraints aren't met.
type DataMessageBodyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataMessageBodyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataMessageBodyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataMessageBodyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataMessageBodyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataMessageBodyValidationError) ErrorName() string { return "DataMessageBodyValidationError" }

// Error satisfies the builtin error interface
func (e DataMessageBodyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataMessageBody.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataMessageBodyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataMessageBodyValidationError{}

// Validate checks the field values on DataMessage with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataMessage) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataMessage with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataMessageMultiError, or
// nil if none found.
func (m *DataMessage) ValidateAll() error {
	return m.validate(true)
}

func (m *DataMessage) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Type

	// no validation rules for RoomId

	// no validation rules for RoomSid

	// no validation rules for To

	if all {
		switch v := interface{}(m.GetBody()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DataMessageValidationError{
					field:  "Body",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DataMessageValidationError{
					field:  "Body",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetBody()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DataMessageValidationError{
				field:  "Body",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.MessageId != nil {
		// no validation rules for MessageId
	}

	if len(errors) > 0 {
		return DataMessageMultiError(errors)
	}

	return nil
}

// DataMessageMultiError is an error wrapping multiple validation errors
// returned by DataMessage.ValidateAll() if the designated constraints aren't met.
type DataMessageMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataMessageMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataMessageMultiError) AllErrors() []error { return m }

// DataMessageValidationError is the validation error returned by
// DataMessage.Validate if the designated constraints aren't met.
type DataMessageValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataMessageValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataMessageValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataMessageValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataMessageValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataMessageValidationError) ErrorName() string { return "DataMessageValidationError" }

// Error satisfies the builtin error interface
func (e DataMessageValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataMessage.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataMessageValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataMessageValidationError{}
